src/transform/src/monotonic.rs

```diff
@@ -116,6 +116,35 @@ impl MonotonicFlag {
                     }
                     result
                 }
+                MirRelationExpr::LetRec { ids, values, body } => {
+                    for id in ids.iter() {
+                        if locals.contains(id) {
+                            panic!("Shadowing of identifier: {:?}", id);
+                        }
+                    }
+                    // Pessimistically assume all bindings are non-monotonic, and
+                    // iteratively attempt to determine that some may be monotonic.
+                    //
+                    // TODO: This is suboptimal as long as there is one method to
+                    // both report and apply monotonicity information. We need to
```

Comment:
This would be good to discuss; it was not obvious to me that it was easy especially if you need to re-run the first traversal multiple times. Like, if you need to pop off a bunch of work from the stack and re-do it all, is that easy or hard in attribute land (have we baked in the idea that it is a tree and post-order is all you need).

Reply:
Yes, the current attribute framework doesn't make this easy, because you cannot easily find the attributes for an arbitrary part of the expression in a subsequent traversal. The current framework just mostly allows for accessing attribute values in the same traversal to nodes you recently visited (e.g., the children of the current node, or digging up an attribute from under a pattern that consists of a few nodes starting at the current node), by looking at some portion of the top of the stack.

What I meant is that it would be great to have a more general attribute framework, allowing for attaching arbitrary attributes to all nodes, and then later easily accessing them in other traversals.